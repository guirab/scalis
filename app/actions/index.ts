"use server";
export async function create(formData: NewAccountType) {
  return (
    await fetch("http://localhost:3000/api/accounts", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }))?.json();
}

export async function getAll() {
  return (
    await fetch("http://localhost:3000/api/accounts")
  )?.json();
}

export async function login(formData: LoginType) {
  return await fetch("http://localhost:3000/api/login", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(formData),
  }).then((res:any) => {return res.json()});
}

export async function transferToSameAcc({action, type, amount, id, from}: UpdateType) {
  return await fetch(`http://localhost:3000/api/account/${id}`, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({action, type, amount, from}),
  }).then((res:any) => {return res.json()});
}

export async function transferToOtherAcc({amount, to, password, id}: TransferOtherAccType) {
  return await fetch(`http://localhost:3000/api/account/transfer/${id}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ amount, to, password }),
  }).then((res:any) => {return res.json()});
}