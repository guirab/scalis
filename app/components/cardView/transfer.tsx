"use client";
import { useState } from "react";

import { TransferSameAcc } from "../transfer/sameAcc";
import { TransferOtherAcc } from "../transfer/otherAcc";

export const TransferView: React.FC<React.HTMLProps<any> & ActionCardType> = ({
  open,
  setOpen,
  ...props
}) => {
  const [isSameAccount, setIsSameAccount] = useState("true");

  const currentView = {
    true: <TransferSameAcc open={open} setOpen={setOpen} />,
    false: <TransferOtherAcc open={open} setOpen={setOpen} />,
  }[isSameAccount];

  return (
    <>
      {open && (
        <div className="block w-full mt-4 space-y-4" {...props}>
          <div className="gap-x-2 mb-4 flex text-sm justify-between">
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="same">
                This account
              </label>
              <input
                data-testid="same-account"
                type="radio"
                id="same"
                value="same"
                checked={isSameAccount === "true"}
                onChange={() => setIsSameAccount("true")}
              />
            </div>
            <div className="flex justify-start items-center">
              <label className="pr-px" htmlFor="other">
                Other account
              </label>
              <input
                data-testid="other-account"
                type="radio"
                id="other"
                value="other"
                checked={isSameAccount === "false"}
                onChange={() => setIsSameAccount("false")}
              />
            </div>
          </div>
          {currentView}
        </div>
      )}
    </>
  );
};
