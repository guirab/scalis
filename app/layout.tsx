import type { Metadata } from "next";
import { Inter } from "next/font/google";
import "./globals.css";

import { AccountsProvider } from "@/store/context";

const inter = Inter({ subsets: ["latin"] });

export const metadata: Metadata = {
  title: "Banking App",
  description: "Banking App for SCALIS",
};

export default function RootLayout({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body className={inter.className}>
        <AccountsProvider>{children}</AccountsProvider>
      </body>
    </html>
  );
}
