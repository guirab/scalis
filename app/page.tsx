import Accounts from "./components/accounts";
import NewAccForm from "./components/newAccForm";

export default function Home() {
  return (
    <main className="flex min-h-screen flex-col items-center p-12 sm:p-24 justify-center sm:justify-start">
      <h1 className="text-5xl">Banking App</h1>
      <div className="w-full flex justify-between flex-col sm:flex-row gap-y-4">
        <NewAccForm />
        <Accounts />
      </div>
    </main>
  );
}
