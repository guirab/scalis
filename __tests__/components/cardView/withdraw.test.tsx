import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import { WithdrawView } from "../../../app/components/cardView/withdraw";
import { afterEach, describe, expect, test, vitest } from "vitest";
import { transferToSameAcc } from "../../../app/actions";

global.fetch = vitest.fn();

function createFetchResponse(data: any) {
  return { json: () => new Promise((resolve) => resolve(data)) };
}

describe("WithdrawView", () => {
  afterEach(() => {
    cleanup();
  });
  test("snapshot", () => {
    const { container } = render(<WithdrawView open setOpen={() => {}} />);
    expect(container).toMatchSnapshot();
  });

  test("updates the type state when the select value changes", () => {
    render(<WithdrawView open setOpen={() => {}} />);

    const select = screen.getByTestId("withdraw-select") as HTMLSelectElement;

    fireEvent.change(select, {
      target: { value: "savings" },
    });
    expect(select.value).toBe("savings");

    fireEvent.change(select, {
      target: { value: "checking" },
    });
    expect(select.value).toBe("checking");
  });

  test("updates the amount state when the input value changes", () => {
    render(<WithdrawView open setOpen={() => {}} />);

    const input = screen.getByLabelText("Amount:") as HTMLInputElement;

    fireEvent.change(input, { target: { value: "100" } });
    expect(input.value).toBe("1.00");
  });

  test("try to withdraw more than the account balance", async () => {
    const response = {
      id: 1,
      username: "test",
      password: "test",
      checking: 1,
      savings: 1,
    };

    vitest.mock("react", async () => {
      return {
        ...(await vitest.importActual("react")),
        useContext: () => ({
          account: { id: 1, username: "test", checking: 1, savings: 1 },
        }),
      };
    });

    render(<WithdrawView open setOpen={vitest.fn()} />);

    const input = screen.getByLabelText("Amount:") as HTMLInputElement;
    fireEvent.change(input, {
      target: { value: "100" },
    });

    fireEvent.click(screen.getByRole("button", { name: "Withdraw" }));

    (fetch as any).mockResolvedValue(createFetchResponse(response));

    const createData = await transferToSameAcc({
      action: "withdraw",
      type: "checking",
      amount: 100,
      id: 1,
    });

    expect(fetch).toHaveBeenCalledWith(
      `http://localhost:3000/api/account/${response.id}`,
      {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          action: "withdraw",
          type: "checking",
          amount: 100,
        }),
      }
    );
    expect(createData).toStrictEqual(response);
  });

  test("calls the onClick function when the button is clicked", async () => {
    const response = {
      id: 1,
      username: "test",
      password: "test",
      checking: 1,
      savings: 1,
    };

    vitest.mock("react", async () => {
      return {
        ...(await vitest.importActual("react")),
        useContext: () => ({
          account: { id: 1, username: "test" },
        }),
      };
    });

    render(<WithdrawView open setOpen={vitest.fn()} />);

    const input = screen.getByLabelText("Amount:") as HTMLInputElement;
    fireEvent.change(input, {
      target: { value: "100" },
    });

    fireEvent.click(screen.getByRole("button", { name: "Withdraw" }));

    (fetch as any).mockResolvedValue(createFetchResponse(response));

    const createData = await transferToSameAcc({
      action: "withdraw",
      type: "checking",
      amount: 100,
      id: 1,
    });

    expect(fetch).toHaveBeenCalledWith(
      `http://localhost:3000/api/account/${response.id}`,
      {
        method: "PUT",
        headers: {
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          action: "withdraw",
          type: "checking",
          amount: 100,
        }),
      }
    );
    expect(createData).toStrictEqual(response);
  });
});
