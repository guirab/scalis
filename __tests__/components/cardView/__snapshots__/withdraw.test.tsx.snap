// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WithdrawView > snapshot 1`] = `
<div>
  <div>
    <label
      for="withdraw"
    >
      From:  
    </label>
    <select
      class="text-black pl-2 outline-none rounded-md w-full"
      data-testid="withdraw-select"
      id="withdraw"
      name="withdraw"
    >
      <option
        value="checking"
      >
        Checking
      </option>
      <option
        value="savings"
      >
        Savings
      </option>
    </select>
    <div>
      <label
        for="currency-withdraw"
      >
        Amount:  
      </label>
      <div
        class="relative w-full"
      >
        <span
          class="absolute text-black left-px"
        >
          $ 
        </span>
        <input
          autocomplete="off"
          class="text-black pl-4 text-black pl-2 appearance-none outline-none rounded-md w-full"
          data-testid="currency-input"
          id="currency-withdraw"
          name="currency-withdraw"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      class="bg-blue-500 text-white px-4 py-2 rounded-md mt-4 w-full"
    >
      Withdraw
    </button>
  </div>
</div>
`;
